std-bits/base/src/main/java/org/enso/base/Array_Builder.java

```diff
@@ -37,21 +41,94 @@ public boolean isEmpty() {
    * @param e the element to add
    */
   public void add(T e) {
-    builder.add(e);
+    if (objectArray != null) {
+        if (size == objectArray.length) {
+            objectArray = Arrays.copyOf(objectArray, size * 2);
+        }
+        objectArray[size++] = e;
+    } else if (primitiveArray instanceof long[] longArray) {
+        if (e instanceof Long l) {
+            if (size == longArray.length) {
+                primitiveArray = longArray = Arrays.copyOf(longArray, size * 2);
+            }
+            longArray[size++] = l;
+        } else {
+            objectArray = new Object[longArray.length];
+            for (int i = 0; i < longArray.length; i++) {
+                objectArray[i] = longArray[i];
+            }
+            primitiveArray = null;
+            add(e);
+        }
+    } else if (primitiveArray instanceof double[] doubleArray) {
+        if (e instanceof Double d) {
+            if (size == doubleArray.length) {
+                primitiveArray = doubleArray = Arrays.copyOf(doubleArray, size * 2);
+            }
+            doubleArray[size++] = d;
+        } else {
+            objectArray = new Object[doubleArray.length];
+            for (int i = 0; i < doubleArray.length; i++) {
+                objectArray[i] = doubleArray[i];
+            }
+            primitiveArray = null;
+            add(e);
+        }
+    } else {
+        assert objectArray == null;
+        assert primitiveArray == null;
+        assert size == 0;
+        switch (e) {
+            case Long l -> {
+                var arr = new long[capacity];
+                arr[0] = l;
+                primitiveArray = arr;
+            }
+            case Double d -> {
+                var arr = new double[capacity];
+                arr[0] = d;
+                primitiveArray = arr;
+            }
+            default -> {
+                var arr = new Object[capacity];
+                arr[0] = e;
+                objectArray = arr;
+            }
+        }
+        size = 1;
+    }
   }
 
   /** Obtains an element from the builder */
-  public T get(int index) {
-    return builder.get(index);
+  public Object get(int index) {
+    if (objectArray != null) {
+        return objectArray[index];
+    } else if (primitiveArray instanceof long[] longArray) {
+        return longArray[index];
+    } else if (primitiveArray instanceof double[] doubleArray) {
+        return doubleArray[index];
+    } else {
+        throw new ArrayIndexOutOfBoundsException();
+    }
   }
 
   /** A temporary workaround to be able to efficiently append an array to `ArrayList`. */
   public void appendTo(List<T> list) {
-    builder.addAll(list);
+    for (T obj : list) {
+        add(obj);
+    }
   }
```

Comment:
I'm wondering if some dispatch tricks we can detect if the `list` is actually backed by an array and then rely on `System.arraycopy` instead of for-iteration (I assume that arraycopy should be quite faster, am I wrong?)

Reply:
`System.arraycopy` doesn't necessarily have to be faster.

Reply:
Oh, ok that is new to me! Could you tell me more? I'm curious :slightly_smiling_face:

Reply:
Here is a simple benchmark: https://github.com/enso-org/enso/compare/develop...JaroslavTulach:enso:jtulach/BenchmarkLoopAndArrayCopy - results on my computer are:
- `int[]` - arraycopy is 5% faster
- `Object[]` arraycopy is 20% faster!
- 'long[]' - almost the same results

Usually it makes little sense to complicate code with `arraycopy` that methods was handy in nineties when Java was slow. Today's compilers can do quite a good job without `arraycopy`.